### ISSUE TITLE
Validation Loss Plateaus only after 2000 steps Pretraining Regardless of LR Schedule (Follow-up to #56)

### ISSUE THREAD

**Issue by jefcoder** [OpenDCAI/DataFlex#57]
Hi everyone,

Following up on my earlier post (#3868) about multi-dataset sampling and learning rate selection .

Thanks so much to this community for all the help — special thanks to @Walnutes. With your input, I've made significant progress, and validation loss across all datasets is now decreasing consistently .

However, I'm now running into a related but more fundamental problem: the **validation loss plateaus hard only after 2000 steps** of continued pretraining, regardless of the learning rate schedule I use .

**Setup**
Hardware: 8×B200s
Run length: 12 hours
Total loss reduction: ~0.0008 over the full run
LR schedules tried: Cosine annealing decay: 5e-6 → 1e-6

**Observed behavior**
Regardless of schedule, the loss curve is heavily front-loaded:

~70% of total reduction happens in the first third of training
~20% in the second third
~10% or less in the final third — essentially flat

<img width="1834" height="681" alt="cosine" src="https://github.com/user-attachments/assets/113ff75d-2611-43d4-9574-5429143f35b7" />
<img width="1808" height="633" alt="image" src="https://github.com/user-attachments/assets/aaeef47a-0622-4607-b41c-15042cda18fb" />

I initially assumed the main culprit was the low LR combined with cosine annealing — especially since LR values below 3e-6 seemed to produce a negligible learning signal, essentially "useless" for training.

So I switched to a constant LR of 5e-6, expecting this to resolve the issue . But I observed the same symptom regardless .
<img width="1838" height="660" alt="constant" src="https://github.com/user-attachments/assets/322a8f3e-8914-4b3b-b3cd-8fb99c8ad0bf" />
<img width="1838" height="633" alt="image" src="https://github.com/user-attachments/assets/21b9d745-215a-4f5f-89cd-56d43b487950" />

I'm now genuinely puzzled as to why the loss only reduces meaningfully at the very beginning of training, then flattens out almost entirely only after 2000 steps.

I'd really appreciate hearing from anyone who's hit a similar wall in large-scale continued pretraining. As before, happy to discuss further via DM, and open to consulting/collaboration if useful.

Thanks in advance!

**Comment by Walnutes** [comment]
Hi @jefcoder,

Sorry for the delayed response. 

Thanks for your patience and the detailed follow-up. Compared with the behavior in [#56](https://github.com/OpenDCAI/DataFlex/issues/56), the qualitative problem has changed: the displayed per-domain losses now generally decrease, suggesting that the earlier under-exposure and forgetting issue has been substantially mitigated. The remaining problem is the rapidly diminishing marginal gain.

The new curves do not show a synchronized failure at exactly step 2000:

- In the `constant-5e-6` run, `eval/weighted_mean_loss` decreases from approximately `1.48365` to `1.48285` by step 2000, and then to about `1.48275` by step 4500. The later gain is small, but not zero.
- The domains flatten at different times. `db-large-2` gains mostly in the early stage, `db-small-3` becomes nearly flat around step 1500, while `db-large-1` continues improving after step 2000.
- The gradient norm remains stable, with no corresponding transition around step 2000.
- `eval/best_loss` appears to be a running minimum, so its long flat segments visually exaggerate the plateau.

The constant-LR run shows that the same front-loaded behavior can occur without cosine decay, making the scheduler unlikely to be the sole cause. However, the cosine and constant runs start from substantially different weighted, mean, and per-domain losses—for example, approximately `1.4406` versus `1.4836` for the weighted loss—and also have different run lengths. They therefore do not yet constitute a controlled comparison of LR schedules.

---

## 1. Optimization dynamics

A stable gradient norm does not imply that the updates remain useful for the weighted validation objective. Validation improvement depends on the alignment between the optimizer-preconditioned training update and the weighted validation gradient, not only on the magnitude of the minibatch gradient.

The constant-LR run keeps a gradient norm near `0.4`, while its weighted-loss improvement per step falls by roughly an order of magnitude. This suggests that gradients have not vanished; rather, later updates provide less useful descent for the target objective. The remaining domain gradients may increasingly cancel one another or improve domains that contribute less to the weighted metric.

This interpretation is also consistent with the cosine run: the unweighted mean loss continues to improve after the weighted loss becomes nearly flat. The model is still learning, but the remaining progress is less aligned with the stated weighted objective. Increasing one global LR would amplify the same aggregate update direction without correcting this misalignment, and could reintroduce the cross-domain interference observed in #56.

---

## 2. Reward–objective alignment

If DataFlex ODM is being used in these runs, it would also be worth confirming that ODM is enabled as intended and that its `warmup_step` is appropriate for the total 3000–4500-step budget. An overly long warm-up may leave too little time for the sampling distribution to adapt.

What's more, ODM uses training loss as an efficient proxy for information gain. Its estimated domain reward can be summarized as:

$$\widehat{R}_i \leftarrow \alpha \widehat{R}_i + (1-\alpha)\frac{c\,L_{i,\mathrm{batch}}^{\mathrm{train}}}{p_i}$$

where `p_i` is the current sampling probability and `c` is the reward scale.

This objective is not necessarily identical to maximizing the remaining weighted validation gain. In late-stage CPT, high loss may reflect intrinsic entropy or difficulty rather than large reducible loss. The displayed domain losses span roughly `0.29–2.14`, so their absolute values are not directly comparable measures of marginal utility. A high-loss but saturated domain may still receive substantial sampling mass while contributing little to the target objective.

For the stated objective, a useful diagnostic is the weighted validation gain per consumed token:

$$s_i(t)=q_i\frac{L_i^{\mathrm{val}}(t-\Delta)-L_i^{\mathrm{val}}(t)}{N_i(t)-N_i(t-\Delta)}$$

Although this is observational rather than causal, comparing it with the evolving sampling proportions can reveal whether additional tokens are being allocated to domains that still provide weighted validation progress.

---

## 3. Controlled comparison

A compact comparison from the same full checkpoint around step 1500–1800 can distinguish the main explanations:

1. Continue the cosine schedule with a fixed mixture to establish the reference trajectory under the current scheduler.
2. Use constant `5e-6` with the same mixture and data order to isolate the effect of LR decay.
3. Use constant `5e-6` with adaptive mixing to test whether reallocating the sampling budget restores weighted validation progress.

If these runs exhibit similar late-stage slopes, short single-domain probes can further distinguish saturation from interference: strong own-domain gains accompanied by regressions on other domains indicate gradient conflict, whereas negligible gains across all domains indicate little remaining learnable signal.

Overall, the evidence does not support a universal 2000-step optimization wall. It is more consistent with asynchronous domain saturation and declining alignment between the training updates and the weighted validation objective.

Hope this helps clarify the issue. If there are more questions or follow-up experiment results, feel free to share them here.
